(INTL-12) Change the POT name from messages to project name.

libbymo opened the pull request:
This PR changes the name of messages.pot, to <project-name>.pot.

This will require manual deletion of this file.  I'm unsure if / where I should put this in the documentation.  

I need to verify with @LarissaLane that it's ok to use the project-name, and not the repo name, as repo-name can be difficult to determine.  Project-name is more easily known, and consistent.

@aperiodic If you could give this a glance, given your work on the CI script and recent tiptoe into the Makefile, I'd appreciate it.

libbymo commented:
Removing the WIP label, mainly as I can't see why project name vs repo name would make a difference. They're pretty obvious, if not the same in most cases.

libbymo commented:
@lutter it'd be nice to have your eyeballs as well...

aperiodic commented on src/leiningen/i18n.clj:
```diff
@@ -79,6 +82,20 @@
+               ; use RegExp's m flag - (?m) - so that ^/$ match start/end of line rather than
+               ; start/end of the entire string
+               #"(?m)^POT_NAME=.*$"
+               (str "POT_NAME=" (:name project) ".pot")))))
```
I think this code and the other single-line replacements (starting at line no. 57) would be easier to understand if the implementation detail of needing to specify the multi-line regex flag were hidden in a utility function that provided a way to replace a line in a multi-line string by providing a line regex and the replacement string and having the utility function take care of the details (you can use the regex Pattern's .toString method to get the form that can be fed to `re-pattern` prepended with flag bit to turn the regex supplied by the caller into a multi-line aware regex).

libbymo replied:
Makes sense.  I'll give it a whirl.

aperiodic commented on src/leiningen/i18n.clj:
```diff
@@ -79,6 +82,20 @@
+  (let [dest (path-join (dev-resources-dir project) "i18n" "bin" "update-pot.sh")
+        pot-script (io/resource "leiningen/i18n/bin/update-pot.sh")]
+    (spit (io/as-file dest)
+          (clojure.string/replace-first
```
I know this was in there already, but we should add `clojure.string` to the requires and alias it, rather than sidestepping the require process by referring to the fully qualified namespace and hoping that a dependency loads that namespace first.

libbymo replied:
Sure thing.

aperiodic commented on src/leiningen/i18n/Makefile:
```diff
@@ -38,9 +42,9 @@ export LOCALES_CLJ_CONTENTS
 i18n: msgfmt
 
 # Update locales/messages.pot
```
This should be changed or removed given that it's no longer called messages.pot

libbymo replied:
Will do.

aperiodic commented on src/leiningen/i18n/bin/update-pot.sh:
```diff
@@ -2,8 +2,10 @@
 
 set -eo pipefail
 
+POT_NAME=""
```
The commit message for when the POT changes are checked in (on line 66) should also be updated to use this here new POT_NAME variable.

libbymo replied:
Good idea.

libbymo commented:
Thanks for the input @aperiodic.  I'll set about updating.

libbymo commented:
@aperiodic I think this is ready.  I think the upgrade path will require:
1) lein i18n init (install / update pot name in script)
2) make i18n (generate the <project-name>.pot
3) remove messages.pot

I think those with CI tasks could skip #2.  But it would probably be good to generate and commit, to verify it's working as expected.  That sound sane? I tested this out in a project.

libbymo commented:
Also let me know, I can squash commits if it makes it cleaner.

libbymo commented:
@aperiodic I squashed commits, and added the utility function, and tests. If there are any test cases that should be removed or added, just let me know. Thanks!

aperiodic commented on src/leiningen/i18n/utils.clj:
```diff
@@ -0,0 +1,11 @@
+(defn replace-first-in-multiline
+  "Replace first instance of regex-match in the passed in string with the
+   replacement. Uses RegExp's m flag (?m) so that ^/$ matches start/end of line."
+  [s match-string replacement]
```
The docstring and the arguments should be updated to reflect that the function accepts a regex rather than a string. It should probably also mention that the regex need only match the beginning of the line (although I'd prefer that this function not add the `.*` before the `$` so the behavior is more apparent at the call site and it's possible for the caller to use a stricter match if they prefer).

aperiodic commented on test/puppetlabs/i18n/utils_test.clj:
```diff
@@ -0,0 +1,30 @@
+    (is (= "\nNew stuff"
+           (utils/replace-first-in-multiline "\nABC-DEF stuff"
+                                             #"ABC-DEF"
+                                             "New stuff")))))
```
just FYI, you can nest `testing` expressions to avoid repetition because nested `testing` expressions get their strings concatenated (with a space between them).

aperiodic commented on README.md:
```diff
@@ -206,18 +208,27 @@ certain test should run, for example, with
+To update the po:
+    msgmerge -U locales/eo.po locales/<project-name>.pot
+
+This uses the contents of the current POT to update msgids in the target po (eo.po).
```
Hmm, when we talked last week, you said that we don't want people to update the PO files, since they should be managed by transifex. If there is cause to update the PO files outside of transifex we should add that functionality back into the Makefile rather than having people call msgmerge themselves.

aperiodic replied:
Oh, I see you have a disclaimer about this up above. That's fine, then.

libbymo replied:
cool. Thanks... yeah it's really I guess a comment for potential `other users` or maybe QA.
